morpheus/utils/nvt/schema_converters.py

```diff
@@ -0,0 +1,430 @@
+
+            # Chain ops together into a compound op
+            node_op = parent_input
+            for op in ops:
+                node_op = node_op >> op
+
+            # Set the op for this node to the compound operator
+            node_op_map[node] = node_op
+
+            # Add our neighbors to the queue
+            neighbors = [n for n in graph.neighbors(node)]
+            for neighbor in neighbors:
+                queue.append(neighbor)
+
+    return visited, node_op_map
+
+
+def coalesce_leaf_nodes(node_op_map, graph, preserve_re):
+    coalesced_workflow = None
+    for node, op in node_op_map.items():
+        neighbors = [n for n in graph.neighbors(node)]
+        # Only add the operators for leaf nodes, or those explicitly preserved
+        if len(neighbors) == 0 or (preserve_re and preserve_re.match(node)):
+            if coalesced_workflow is None:
+                coalesced_workflow = op
+            else:
+                coalesced_workflow = coalesced_workflow + op
+
+    return coalesced_workflow
+
+
+def coalesce_ops(graph, ci_map, preserve_re=None):
+    root_nodes = [node for node, in_degree in graph.in_degree() if in_degree == 0]
+    visited, node_op_map = bfs_traversal_with_op_map(graph, ci_map, root_nodes)
+    coalesced_workflow = coalesce_leaf_nodes(node_op_map, graph, preserve_re=preserve_re)
+
+    return coalesced_workflow
+
+
+def dataframe_input_schema_to_nvt_workflow(input_schema: DataFrameInputSchema, visualize=False) -> nvt.Workflow:
+    """
+    Converts an `input_schema` to a `nvt.Workflow` object
+
+    First we aggregate all preprocessing steps, which we assume are independent of each other and can be run in
+    parallel.
+
+    Next we aggregate all column operations, which we assume are independent of each other and can be run in parallel
+    and pass them the updated schema from the preprocessing steps.
+    """
+
+    if (input_schema is None or len(input_schema.column_info) == 0):
+        raise ValueError("Input schema is empty")
+
+    # Try to guess which output columns we'll produce
+    json_output_cols = resolve_json_output_columns(input_schema)
+
+    json_cols = input_schema.json_columns
+    column_info_objects = [ci for ci in input_schema.column_info]
+    if (json_cols is not None and len(json_cols) > 0):
+        column_info_objects.append(
+            JSONFlattenInfo(
+                input_col_names=[c for c in json_cols],
+                # output_col_names=[name for name, _ in json_output_cols],
+                output_col_names=json_output_cols,
+                dtype="str",
+                name="json_info"))
+
+    column_info_map = {ci.name: ci for ci in column_info_objects}
+
+    graph = build_nx_dependency_graph(column_info_objects)
+
+    # Uncomment to print the computed: dependency layout
+    # from matplotlib import pyplot as plt
+    # pos = graphviz_layout(graph, prog='neato')
+    # nx.draw(graph, pos, with_labels=True, font_weight='bold')
+    # plt.show()
```

Comment:
Maybe switch this to work off an environment variable to print the debug info when needed?